File: backend/Controllers/ExamController.cs
using System.Text.Json;
using LanguageCenter.Api.Infrastructure;
using LanguageCenter.Api.Services;
using Microsoft.AspNetCore.Mvc;

namespace LanguageCenter.Api.Controllers;

[Route("api/{tenantSlug}")]
public sealed class ExamController : ApiControllerBase
{
    private readonly ExamService _exams;
    public ExamController(PermissionService permissions, ExamService exams) : base(permissions) => _exams = exams;

    [HttpGet("exam-formats")]
    public async Task<IActionResult> Formats() { var (_, _, e) = await RequireContextAsync(); if (e is not null) return e; return Ok(new { formats = _exams.GetFormats() }); }
    [HttpPost("exam-media")]
    [RequestSizeLimit(104857600)]
    public async Task<IActionResult> UploadMedia([FromForm] IFormFile file) { var (_, _, e) = await RequireContextAsync(true, "admin", "staff", "teacher"); if (e is not null) return e; return Ok(await _exams.UploadMediaAsync(file)); }
    [HttpGet("exam-question-groups")]
    public async Task<IActionResult> Groups() { var (t, _, e) = await RequireContextAsync(true, "admin", "staff", "teacher"); if (e is not null) return e; return Ok(await _exams.ListGroupsAsync(t!.Id, Request.Query)); }
    [HttpGet("exam-questions")]
    public async Task<IActionResult> Questions() { var (t, _, e) = await RequireContextAsync(true, "admin", "staff", "teacher"); if (e is not null) return e; return Ok(await _exams.ListQuestionsAsync(t!.Id, Request.Query)); }
    [HttpGet("exam-questions/{id:int}")]
    public async Task<IActionResult> Question(int id) { var (t, _, e) = await RequireContextAsync(true, "admin", "staff", "teacher"); if (e is not null) return e; var r = await _exams.GetQuestionAsync(t!.Id, id); return r is null ? NotFound(new { message = "Question not found" }) : Ok(r); }
    [HttpPost("exam-questions")]
    public async Task<IActionResult> CreateQuestion([FromBody] JsonElement body) { var (t, u, e) = await RequireContextAsync(true, "admin", "staff", "teacher"); if (e is not null) return e; return Created("", await _exams.CreateQuestionAsync(t!.Id, u!.Id, body)); }
    [HttpPut("exam-questions/{id:int}")]
    public async Task<IActionResult> UpdateQuestion(int id, [FromBody] JsonElement body) { var (t, _, e) = await RequireContextAsync(true, "admin", "staff", "teacher"); if (e is not null) return e; var r = await _exams.UpdateQuestionAsync(t!.Id, id, body); return r is null ? NotFound(new { message = "Question not found" }) : Ok(r); }
    [HttpDelete("exam-questions/{id:int}")]
    public async Task<IActionResult> DeleteQuestion(int id) { var (t, _, e) = await RequireContextAsync(true, "admin"); if (e is not null) return e; await _exams.DeleteQuestionAsync(t!.Id, id); return Ok(new { message = "Deleted successfully" }); }

    [HttpGet("exam-sets")]
    public async Task<IActionResult> ExamSets() { var (t, _, e) = await RequireContextAsync(true, "admin", "staff", "teacher"); if (e is not null) return e; return Ok(await _exams.ListExamSetsAsync(t!.Id, Request.Query)); }
    [HttpGet("exam-sets/{id:int}")]
    public async Task<IActionResult> ExamSet(int id) { var (t, _, e) = await RequireContextAsync(true, "admin", "staff", "teacher"); if (e is not null) return e; var r = await _exams.GetExamSetAsync(t!.Id, id); return r is null ? NotFound(new { message = "Exam set not found" }) : Ok(r); }
    [HttpPost("exam-sets")]
    public async Task<IActionResult> CreateExamSet([FromBody] JsonElement body) { var (t, u, e) = await RequireContextAsync(true, "admin", "staff"); if (e is not null) return e; return Created("", await _exams.CreateExamSetAsync(t!.Id, u!.Id, body)); }
    [HttpPut("exam-sets/{id:int}")]
    public async Task<IActionResult> UpdateExamSet(int id, [FromBody] JsonElement body) { var (t, _, e) = await RequireContextAsync(true, "admin", "staff"); if (e is not null) return e; var r = await _exams.UpdateExamSetAsync(t!.Id, id, body); return r is null ? NotFound(new { message = "Exam set not found" }) : Ok(r); }
    [HttpPost("exam-sets/generate")]
    public async Task<IActionResult> GenerateExamSet([FromBody] JsonElement body) { var (t, u, e) = await RequireContextAsync(true, "admin", "staff"); if (e is not null) return e; return Created("", await _exams.GenerateExamSetAsync(t!.Id, u!.Id, body)); }
    [HttpDelete("exam-sets/{id:int}")]
    public async Task<IActionResult> DeleteExamSet(int id) { var (t, _, e) = await RequireContextAsync(true, "admin"); if (e is not null) return e; await _exams.DeleteExamSetAsync(t!.Id, id); return Ok(new { message = "Deleted successfully" }); }

    [HttpGet("mock-exams")]
    public async Task<IActionResult> MockExams() { var (t, _, e) = await RequireContextAsync(); if (e is not null) return e; return Ok(await _exams.ListMockExamsAsync(t!.Id, Request.Query)); }
    [HttpGet("mock-exams/{id:int}")]
    public async Task<IActionResult> MockExam(int id) { var (t, _, e) = await RequireContextAsync(); if (e is not null) return e; var r = await _exams.GetMockExamAsync(t!.Id, id); return r is null ? NotFound(new { message = "Mock exam not found" }) : Ok(r); }
    [HttpPost("mock-exams")]
    public async Task<IActionResult> CreateMockExam([FromBody] JsonElement body) { var (t, _, e) = await RequireContextAsync(true, "admin", "staff"); if (e is not null) return e; return Created("", await _exams.CreateMockExamAsync(t!.Id, body)); }
    [HttpPut("mock-exams/{id:int}")]
    public async Task<IActionResult> UpdateMockExam(int id, [FromBody] JsonElement body) { var (t, _, e) = await RequireContextAsync(true, "admin", "staff"); if (e is not null) return e; var r = await _exams.UpdateMockExamAsync(t!.Id, id, body); return r is null ? NotFound(new { message = "Mock exam not found" }) : Ok(r); }
    [HttpDelete("mock-exams/{id:int}")]
    public async Task<IActionResult> DeleteMockExam(int id) { var (t, _, e) = await RequireContextAsync(true, "admin"); if (e is not null) return e; await _exams.DeleteMockExamAsync(t!.Id, id); return Ok(new { message = "Deleted successfully" }); }
    [HttpPost("mock-exam-students/{mockExamStudentId:int}/start")]
    public async Task<IActionResult> Start(int mockExamStudentId) { var (t, _, e) = await RequireContextAsync(); if (e is not null) return e; return Ok(await _exams.StartAttemptAsync(t!.Id, mockExamStudentId)); }
    [HttpPost("mock-exam-students/{mockExamStudentId:int}/reset-in-progress")]
    public async Task<IActionResult> Reset(int mockExamStudentId) { var (t, _, e) = await RequireContextAsync(); if (e is not null) return e; return Ok(await _exams.ResetInProgressAsync(t!.Id, mockExamStudentId)); }
    [HttpPost("mock-exam-students/{mockExamStudentId:int}/save-answer")]
    public async Task<IActionResult> SaveAnswer(int mockExamStudentId, [FromBody] JsonElement body) { var (t, _, e) = await RequireContextAsync(); if (e is not null) return e; return Ok(await _exams.SaveAnswerAsync(t!.Id, mockExamStudentId, body)); }
    [HttpPost("mock-exam-students/{mockExamStudentId:int}/recording")]
    [RequestSizeLimit(104857600)]
    public async Task<IActionResult> Recording(int mockExamStudentId, [FromForm] IFormFile file) { var (t, _, e) = await RequireContextAsync(); if (e is not null) return e; return Ok(await _exams.UploadMediaAsync(file)); }
    [HttpPost("mock-exam-students/{mockExamStudentId:int}/submit")]
    public async Task<IActionResult> Submit(int mockExamStudentId, [FromBody] JsonElement body) { var (t, _, e) = await RequireContextAsync(); if (e is not null) return e; return Ok(await _exams.SubmitAsync(t!.Id, mockExamStudentId, body)); }
    [HttpPatch("mock-exam-students/{mockExamStudentId:int}/grade")]
    public async Task<IActionResult> Grade(int mockExamStudentId, [FromBody] JsonElement body) { var (t, _, e) = await RequireContextAsync(true, "admin", "staff", "teacher"); if (e is not null) return e; return Ok(await _exams.GradeAsync(t!.Id, mockExamStudentId, body)); }
}

File: backend/src/controllers/exam.controller.js
const path = require('path');
const fs = require('fs');
const examService = require('../services/exam.service');

class ExamController {
  async getFormats(req, res, next) { try { res.json({ formats: examService.getFormats() }); } catch (err) { next(err); } }
  async listQuestionGroups(req, res, next) { try { res.json(await examService.listQuestionGroups(req.tenant.id, req.query)); } catch (err) { next(err); } }
  async listQuestions(req, res, next) { try { res.json(await examService.listQuestions(req.tenant.id, req.query)); } catch (err) { next(err); } }
  async getQuestion(req, res, next) { try { const item = await examService.getQuestion(req.tenant.id, parseInt(req.params.id)); if (!item) return res.status(404).json({ message: 'Question not found' }); res.json(item); } catch (err) { next(err); } }
  async createQuestion(req, res, next) { try { res.status(201).json(await examService.createQuestion(req.tenant.id, req.user?.id, req.body)); } catch (err) { next(err); } }
  async updateQuestion(req, res, next) { try { const item = await examService.updateQuestion(req.tenant.id, parseInt(req.params.id), { ...req.body, updatedBy: req.user?.id }); if (!item) return res.status(404).json({ message: 'Question not found' }); res.json(item); } catch (err) { next(err); } }
  async deleteQuestion(req, res, next) { try { await examService.deleteQuestion(req.tenant.id, parseInt(req.params.id)); res.json({ message: 'Deleted successfully' }); } catch (err) { next(err); } }

  async listExamSets(req, res, next) { try { res.json(await examService.listExamSets(req.tenant.id, req.query)); } catch (err) { next(err); } }
  async getExamSet(req, res, next) { try { const item = await examService.getExamSet(req.tenant.id, parseInt(req.params.id)); if (!item) return res.status(404).json({ message: 'Exam set not found' }); res.json(item); } catch (err) { next(err); } }
  async createExamSet(req, res, next) { try { res.status(201).json(await examService.createExamSet(req.tenant.id, req.user?.id, req.body)); } catch (err) { next(err); } }
  async updateExamSet(req, res, next) { try { const item = await examService.updateExamSet(req.tenant.id, parseInt(req.params.id), req.body); if (!item) return res.status(404).json({ message: 'Exam set not found' }); res.json(item); } catch (err) { next(err); } }
  async generateExamSet(req, res, next) { try { res.status(201).json(await examService.generateExamSet(req.tenant.id, req.user?.id, req.body)); } catch (err) { next(err); } }
  async deleteExamSet(req, res, next) { try { await examService.deleteExamSet(req.tenant.id, parseInt(req.params.id)); res.json({ message: 'Deleted successfully' }); } catch (err) { next(err); } }

  async listMockExams(req, res, next) { try { res.json(await examService.listMockExams(req.tenant.id, req.query, req.user)); } catch (err) { next(err); } }
  async getMockExam(req, res, next) { try { const item = await examService.getMockExam(req.tenant.id, parseInt(req.params.id), req.user); if (!item) return res.status(404).json({ message: 'Mock exam not found' }); res.json(item); } catch (err) { next(err); } }
  async createMockExam(req, res, next) { try { res.status(201).json(await examService.createMockExam(req.tenant.id, req.user?.id, req.body)); } catch (err) { next(err); } }
  async updateMockExam(req, res, next) { try { const item = await examService.updateMockExam(req.tenant.id, parseInt(req.params.id), req.body); if (!item) return res.status(404).json({ message: 'Mock exam not found' }); res.json(item); } catch (err) { next(err); } }
  async deleteMockExam(req, res, next) { try { await examService.deleteMockExam(req.tenant.id, parseInt(req.params.id)); res.json({ message: 'Deleted successfully' }); } catch (err) { next(err); } }
  async startAttempt(req, res, next) { try { res.json(await examService.startAttempt(req.tenant.id, parseInt(req.params.mockExamStudentId))); } catch (err) { next(err); } }
  async resetInProgressAttempt(req, res, next) { try { res.json(await examService.resetInProgressAttempt(req.tenant.id, parseInt(req.params.mockExamStudentId))); } catch (err) { next(err); } }
  async saveAnswer(req, res, next) { try { res.json(await examService.saveAnswer(req.tenant.id, parseInt(req.params.mockExamStudentId), req.body.answer || req.body)); } catch (err) { next(err); } }
  async saveAnswersBatch(req, res, next) {
    try {
      const tenantId = req.tenant.id;
      const mockExamStudentId = parseInt(req.params.mockExamStudentId);
      const answers = Array.isArray(req.body.answers) ? req.body.answers : [req.body.answer || req.body];
      const results = [];
      for (const answer of answers) {
        const saved = await examService.saveAnswer(tenantId, mockExamStudentId, answer);
        if (saved) results.push(saved);
      }
      res.json({ saved: results.length, results });
    } catch (err) { next(err); }
  }
  async submitAnswers(req, res, next) { try { res.json(await examService.submitAnswers(req.tenant.id, parseInt(req.params.mockExamStudentId), req.body.answers)); } catch (err) { next(err); } }
  async gradeStudent(req, res, next) { try { res.json(await examService.gradeStudent(req.tenant.id, parseInt(req.params.mockExamStudentId), req.body, req.user?.id, req.user)); } catch (err) { next(err); } }

  async uploadSubmissionRecording(req, res, next) {
    try {
      if (!req.file) return res.status(400).json({ message: 'No file uploaded' });
      if (!/^audio\//.test(req.file.mimetype)) return res.status(400).json({ message: 'Chỉ cho phép upload file audio cho phần Speaking' });
      const publicUrl = `/uploads/exams/${req.file.filename}`;
      res.status(201).json({
        url: publicUrl,
        filename: req.file.filename,
        originalName: req.file.originalname,
        mimeType: req.file.mimetype,
        size: req.file.size,
      });
    } catch (err) { next(err); }
  }

  async uploadMedia(req, res, next) {
    try {
      if (!req.file) return res.status(400).json({ message: 'No file uploaded' });
      const publicUrl = `/uploads/exams/${req.file.filename}`;
      res.status(201).json({ url: publicUrl, filename: req.file.filename, originalName: req.file.originalname, mimeType: req.file.mimetype, size: req.file.size });
    } catch (err) { next(err); }
  }
}

module.exports = new ExamController();
